linode/linode_instance_helpers.go

```diff
@@ -702,25 +712,53 @@ func changeInstanceType(client *linodego.Client, instance *linodego.Instance, ta
 			return fmt.Errorf("Error waiting for instance %d readiness: %s", instance.ID, err)
 		}
 	}
+	// We have to wait through the resize process because if we issue jobs before the complete process is complete, the API
+	// Will raise an error.
 
-	if err := client.ResizeInstance(context.Background(), instance.ID, targetType); err != nil {
+	// Issue the resize job
+	if err := client.ResizeInstance(context.Background(), instance.ID, resizeOpts); err != nil {
 		return fmt.Errorf("Error resizing instance %d: %s", instance.ID, err)
 	}
 
-	_, err := client.WaitForEventFinished(context.Background(), instance.ID, linodego.EntityLinode, linodego.ActionLinodeResize, *instance.Created, int(d.Timeout(schema.TimeoutUpdate).Seconds()))
-	if err != nil {
-		return fmt.Errorf("Error waiting for instance %d to finish resizing: %s", instance.ID, err)
+	// wait for instance to begin resizing after issuing resize job
+	if _, err := client.WaitForInstanceStatus(context.Background(), instance.ID, linodego.InstanceResizing, int(d.Timeout(schema.TimeoutUpdate).Seconds())); err != nil {
+		return fmt.Errorf("Error waiting for instance %d to enter resizing state: %s", instance.ID, err)
 	}
 
+	// Wait for instance status to go offline
+	if _, err := client.WaitForInstanceStatus(context.Background(), instance.ID, linodego.InstanceOffline, int(d.Timeout(schema.TimeoutUpdate).Seconds())); err != nil {
+		return fmt.Errorf("Error waiting for instance %d to enter offline state: %s", instance.ID, err)
+	}
+
+	// Wait for instance status to go online if necessary
+	if waitForOnline == true {
+		if _, err := client.WaitForInstanceStatus(context.Background(), instance.ID, linodego.InstanceRunning, int(d.Timeout(schema.TimeoutUpdate).Seconds())); err != nil {
+			return fmt.Errorf("Error waiting for instance %d to enter online state: %s", instance.ID, err)
+		}
+	}
 	return nil
 }
 
 func changeInstanceDiskSize(client *linodego.Client, instance linodego.Instance, disk linodego.InstanceDisk, targetSize int, d *schema.ResourceData) error {
 	if instance.Specs.Disk > targetSize {
-		client.ResizeInstanceDisk(context.Background(), instance.ID, disk.ID, targetSize)
+		if instance.Status == linodego.InstanceOffline || instance.Status == linodego.InstanceShuttingDown {
```

Comment:
This checks to see if `instance.Status == linodego.InstanceOffline`, but the next line waits for the status to become `linodego.InstanceOffline`. That doesn't seem right; maybe make this a `switch` like so:
```go
switch instance.Status {
case linodego.InstanceShuttingDown:
	if _, err := client.WaitForInstanceStatus(context.Background(), instance.ID, linodego.InstanceOffline, int(d.Timeout(schema.TimeoutUpdate).Seconds())); err != nil {
		return fmt.Errorf("Error waiting for instance %d to go offline: %s", instance.ID, err)
	}
case linodego.InstanceOffline:
	continue
default:
	if errShutdown := client.ShutdownInstance(context.Background(), instance.ID); errReturn != nil {
		return errShutdown
	}
}
```

Reply:
Doing this.